Add package-wide logging, resolve #251

widdowquinn opened the pull request:
This PR adds the module `sns_logger.py` which can be used to provide a single centralised package-level logger, as indicated in issue [#251](https://github.com/ScottishCovidResponse/SCRCIssueTracking/issues/251)

To use the logger in a module file:

```python
from simple_network_sim import sns_logger

logger = sns_logger.logger
```

The first import of the module instantiates a `logging.Logger` object that is then available as `sns_logger.logger`. Calls to `logger.info()`, `logger.debug()`, `logger.warning() `etc. from any importing module are caught by the same logger. By default, this logs to STDERR for WARNING and above.

The `sns_logger.setup_logger()` function is available to process `argparse.Namespace` objects obtained from parsing the CLI and modify logger configuration. Currently, this can add a logfile handler, set the STDERR handler to INFO, or set both logfile and STDERR handlers to DEBUG.

The logger is used with the CLI in `sampleUseOfModel.main()` and in the `common.generateMeanPlot()` function to replace a debugging print statement.

This PR also adds the CLI options `-v`/`--verbose`, `--debug` and `-l`/`--logfile`.

aflag commented on simple_network_sim/sns_logger.py:
```diff
@@ -0,0 +1,108 @@
+    logging.config.dictConfig(logconf)
+
+# Provide a package-level logging.Logger object on import
+logger = logging.getLogger(__package__)
```
By not calling this inside each module we won't have the module name in the log line anymore. I don't see what's the issue we are preventing by creating this here.

widdowquinn replied:
It wasn't clear to me that having the module name in the log line was a requirement.

Calling `logging.getLogger(__name__)` creates a new logger with its own scope, that also requires its own configuration. This brings extra overhead in coding, but doesn't give extra functionality. This makes adding extra handlers more difficult, too.

All modules in a package are related at the package level, and a common logger needs only to be configured once to write to the appropriate stream(s) (e.g. a syslog server, log file, STDOUT, etc.). All modules contribute automatically to the same stream(s), which may be set to distinct reporting levels.

If there is a need to identify the module name for information/help with debuggin, this can still be done explicitly in the log message, e.g. `logger.info("Failed in module %s.", __name__), or by including the exception info with `exc_info=True`).

aflag replied:
We don't need to configure each logger individualy, we can just configure the root logger. All the other loggers will inherit from it. By calling `logging.getLogger(__name__)` in each module we get hierarchical logging for free. Python automatically parses `a.b.c` to create the hierarchical logging. This has benefits like changing the log level of each package or module individually if we wanted (and also the module name in the logging line, which I find nice). I think the main reason I'm questioning it is that using `logging.getLogger` is sort of the recommended way to do it in the python docs, I was just trying to understand better the issue with it. Sorry if I'm being annoying, I don't think this is super important.

widdowquinn replied:
Don't worry - it's not annoying. It's helping me rationalise and understand things better (but it's getting late).

aflag commented on simple_network_sim/sns_logger.py:
```diff
@@ -0,0 +1,108 @@
+
+# Provide a package-level logging.Logger object on import
+logger = logging.getLogger(__package__)
+setup_logger()
```
I suggest just calling this in the main module explicitly. It's not really that important in our case (we don't intend for people to use any of the code as a library anyway), but it can be a bit surprising.

widdowquinn replied:
Having this construction enables use of the common logger in each of the modules via an import. An alternative (and harder to maintain) mechanism might be to pass the `logger` explicitly as an argument to each and every function you want to use the logger for.

I understood the intent behind the logger to be to enable support for debugging and (implicitly) to report on issues/status when running via a CLI, as in `sampleUseOfModel.py`. Instantiating the logger only in `main()` doesn't provide support for debugging when modules are imported. This approach gives the same functionality as instantiating in `main()`, and supports `logger.debug()` messages - as opposed to `print()` statements (which I think came up in discussion).

aflag replied:
Hm, I think it's just a bit of pet peeve of mine, because I've dealt with libraries setting up the logger config in past and that and that was really annoying. But I think you're right. For this project, people are probably more interested in being able to import modules from the python shell and having it printing stuff rather than making the code into a library.

aflag commented on simple_network_sim/sns_logger.py:
```diff
@@ -0,0 +1,108 @@
+        "loggers": {
+            __package__: {
+                "handlers": ["stderr"],
+                "level": "DEBUG",
```
I feel like INFO should be the default. For instance, the parameters are printed in the INFO level, and you probably want to see that after each run. It's common for schedulers to run a command and then give you access to the stdout/stderr. But even when you just run locally, it'll probably be useful to be able to go back in history and what all the parameters were.

aflag commented on simple_network_sim/sns_logger.py:
```diff
@@ -0,0 +1,108 @@
+            },
+        "handlers": {
+            "stderr": {
+                "level": "WARNING",
```
I feel like INFO should be the default. For instance, the parameters are printed in the INFO level, and you probably want to see that after each run. It's common for schedulers to run a command and then give you access to the stdout/stderr. But even when you just run locally, it'll probably be useful to be able to go back in history and what all the parameters were.

widdowquinn replied:
I don't think this gets in the way of what you're asking for.

`DEBUG` is the lowest reportable state for the logger, but the handlers defined in `"handlers"` override this. Slightly above that in the code (line 61) the lowest reported status for the default loghandler that writes to STDERR* is defined: this is `WARNING` (by default). That `WARNING` status is overridden if `verbose` or `debug` options are selected in the CLI (lines 96-101).

The overall behaviour is that:

- by default, only `WARNING` and above are written to STDERR
- if `verbose` is flagged, `INFO` and above are written to STDERR
- if `debug` is flagged, `DEBUG` and above are written to STDERR
- if `logfile` is provided, `INFO` and above are written to a file whose name is provided (this is `DEBUG` and above if `debug` is flagged).

This provides some flexibility in providing user messages through the logger:

- fatal problems can be flagged with `logger.error()` (always visible)
- worrying problems that should always be seen can be flagged with `logger.warning()` (always visible)
- useful but nonessential messages can be controlled with `verbose` (default: off)
- messages only useful for monitoring the simulation/run can be controlled with `debug` (default: off).

I'm not wedded to default settings, but I hoped you'd find this flexibility useful.

*The default loghandler writes to STDERR because this is where I'd expect error/information messages to be - STDOUT tends to be reserved for output you might pipe elsewhere.

aflag replied:
I understand. My suggestion was that INFO was written by default (without the need of passing verbose). I had initially commented in the wrong line, that may have caused some confusion.

widdowquinn commented:
You may find it helpful to explore the logger's behaviour with a couple of CLI runs:

```bash
python -m simple_network_sim.sampleUseOfModel -v random afilename.pdf
python -m simple_network_sim.sampleUseOfModel -v -l testlog.log random afilename.pdf
python -m simple_network_sim.sampleUseOfModel -v -l testlog.log random afilename.pdf
python -m simple_network_sim.sampleUseOfModel --debug -l testlog.log random afilename.pdf
```

widdowquinn commented:
Changes as discussed on Zulip:

- moved `setup_logger()` into `sampleUseOfModel.py`, removed `sns_logger.py`
- module-level logging reverted to `logging.getLogger(__name__)` calls; these pick up the package-level configuration when `sampleUseOfModel.py` is `__main__`.
